src/traces/bar/plot.js

```diff
@@ -108,12 +115,274 @@ module.exports = function plot(gd, plotinfo, cdbar) {
                         y0 = fixpx(y0, y1);
                         y1 = fixpx(y1, y0);
                     }
-                    d3.select(this).attr('d',
+
+                    // append bar path and text
+                    var bar = d3.select(this);
+
+                    bar.append('path').attr('d',
                         'M' + x0 + ',' + y0 + 'V' + y1 + 'H' + x1 + 'V' + y0 + 'Z');
+
+                    appendBarText(gd, bar, d, i, x0, x1, y0, y1);
                 });
         });
 
     // error bars are on the top
     bartraces.call(ErrorBars.plot, plotinfo);
 
 };
+
+function appendBarText(gd, bar, calcTrace, i, x0, x1, y0, y1) {
+    var trace = calcTrace[0].trace;
+
+    // get bar text
+    var traceText = trace.text;
+    if(!traceText) return;
+
+    var text;
+    if(Array.isArray(traceText)) {
+        if(i >= traceText.length) return;
+        text = traceText[i];
+    }
+    else {
+        text = traceText;
+    }
+
+    // get text position
+    var traceTextPosition = trace.textposition,
+        textPosition;
+    if(Array.isArray(traceTextPosition)) {
+        if(i >= traceTextPosition.length) return;
+        textPosition = traceTextPosition[i];
+    }
+    else {
+        textPosition = traceTextPosition;
+    }
+
+    if(textPosition === 'none') return;
+
+    var barWidth = Math.abs(x1 - x0),
+        barHeight = Math.abs(y1 - y0),
+        barIsTooSmall = (barWidth < 8 || barHeight < 8),
+
+        barmode = gd._fullLayout.barmode,
+        inStackMode = (barmode === 'stack'),
+        inRelativeMode = (barmode === 'relative'),
+        inStackOrRelativeMode = inStackMode || inRelativeMode,
+
+        calcBar = calcTrace[i],
+        isOutmostBar = !inStackOrRelativeMode || calcBar._outmost;
+
+    if(textPosition === 'auto') {
+        textPosition = (barIsTooSmall && isOutmostBar) ? 'outside' : 'inside';
+    }
+
+    if(textPosition === 'outside') {
+        if(!isOutmostBar) textPosition = 'inside';
+    }
+
+    if(textPosition === 'inside') {
+        if(barIsTooSmall) return;
+    }
```

Comment:
For pies we did this a bit differently - in `'inside'` mode we would always draw the text inside the slice, and shrink it if need be so it would fit (no matter how small the space is). And in `'auto'` mode we would first draw it inside, then if it had to be scaled down to fit inside we'd move it outside.

This algorithm should be easier with a rectangular bar than with a wedge-shaped pie slice - do you want to give it a try?